tool/tbot/main.go

```diff
@@ -0,0 +1,631 @@
+		utils.FatalError(err)
+		trace.DebugReport(err)
+	}
+}
+
+func Run(args []string) error {
+	var cf config.CLIConf
+	utils.InitLogger(utils.LoggingForDaemon, logrus.InfoLevel)
+
+	app := utils.InitCLIParser("tbot", "tbot: Teleport Credential Bot").Interspersed(false)
+	app.Flag("debug", "Verbose logging to stdout").Short('d').BoolVar(&cf.Debug)
+	app.Flag("config", "tbot.yaml path").Short('c').StringVar(&cf.ConfigPath)
+
+	startCmd := app.Command("start", "Starts the renewal bot, writing certificates to the data dir at a set interval.")
+	startCmd.Flag("auth-server", "Specify the Teleport auth server host").Short('a').Envar(authServerEnvVar).StringVar(&cf.AuthServer)
+	startCmd.Flag("token", "A bot join token, if attempting to onboard a new bot; used on first connect.").Envar(tokenEnvVar).StringVar(&cf.Token)
+	startCmd.Flag("ca-pin", "A repeatable auth server CA hash to pin; used on first connect.").StringsVar(&cf.CAPins)
+	startCmd.Flag("data-dir", "Directory to store internal bot data.").StringVar(&cf.DataDir)
+	startCmd.Flag("destination-dir", "Directory to write generated certificates").StringVar(&cf.DestinationDir)
+	startCmd.Flag("certificate-ttl", "TTL of generated certificates").Default("60m").DurationVar(&cf.CertificateTTL)
+	startCmd.Flag("renew-interval", "Interval at which certificates are renewed; must be less than the certificate TTL.").DurationVar(&cf.RenewInterval)
+
+	configCmd := app.Command("config", "Parse and dump a config file").Hidden()
+
+	initCmd := app.Command("init", "Initialize a certificate destination directory.")
+
+	watchCmd := app.Command("watch", "Watch a destination directory for changes.")
+
+	command, err := app.Parse(args)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// While in debug mode, send logs to stdout.
+	if cf.Debug {
+		utils.InitLogger(utils.LoggingForDaemon, logrus.DebugLevel)
+	}
+
+	botConfig, err := config.FromCLIConf(&cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	switch command {
+	case startCmd.FullCommand():
+		err = onStart(botConfig)
+	case configCmd.FullCommand():
+		err = onConfig(botConfig)
+	case initCmd.FullCommand():
+		err = onInit(botConfig)
+	case watchCmd.FullCommand():
+		err = onWatch(botConfig)
+	default:
+		// This should only happen when there's a missing switch case above.
+		err = trace.BadParameter("command %q not configured", command)
+	}
+
+	return err
+}
+
+func onConfig(botConfig *config.BotConfig) error {
+	pretty.Println(botConfig)
+
+	return nil
+}
+
+func onInit(botConfig *config.BotConfig) error {
+	return trace.NotImplemented("init not yet implemented")
+}
+
+func onWatch(botConfig *config.BotConfig) error {
+	return trace.NotImplemented("watch not yet implemented")
+}
+
+func onStart(botConfig *config.BotConfig) error {
+	// Start by loading the bot's primary destination.
+	dest, err := botConfig.Storage.GetDestination()
+	if err != nil {
+		return trace.WrapWithMessage(err, "could not read bot storage destination from config")
```

Comment:
Just curious why not just trace.Wrap(err, "xxx"). Same outcome but a bit shorter.

Reply:
Oh, good to know! I didn't realize it supported formatting args. I'll replace it everywhere.